Module info

petebacondarwin opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

feature

**What is the current behavior? (You can also link to an open issue here)**

No way to find information about the currently loaded modules

**What is the new behavior (if this is a feature change)?**

Two new features:
- a method on `Module` (`info()`) that lets developers add arbitrary info about their modules.
- a property on `$injector` (`modules`) that lets developers access the module objects that were loaded into the injector at bootstrap.

This second item could also be useful to @ocombe for his ocLazyLoad project.

**Does this PR introduce a breaking change?**

No

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Other information**:

Replaces the previous #12465 PR

ocombe commented:
It's happening !!!!

![OMG](https://cloud.githubusercontent.com/assets/265378/19191903/946dcb3c-8ca4-11e6-9840-23de774d77e6.gif)

![Is this real life?](https://cloud.githubusercontent.com/assets/265378/19191961/c64b1312-8ca4-11e6-9d4b-bb8ae7afa35b.gif)

gkalpak commented on src/loader.js:
````diff
@@ -115,6 +118,39 @@ function setupModuleLoader(window) {
+           * {@link $injector#modules `$injector.modules`} property:
+           *
+           * ```js
+           * var version = $injector.modules['myModule'].info().version;
````
Wouldn't `angular.module('myModule').info()` also work?

petebacondarwin replied:
Yes, but in that case you do know, necessarily if the module was
a) loaded in the first place
b) added to the injector
If it was not loaded then your line of code would throw an exception

gkalpak replied:
I mean use it in place of `$injector.modules['myModule'].info()` - i.e. at "runtime" - so it should have been added to the injector.
I was mainly curious basically. `$injector.modules['myModule']` works for me :smiley:

--
Hm...since we brought it up, `$injector` is kind of ambiguous, since it may refer to the `providerInjector` (which doesn't know about `modules`) or the `instanceInjector` (which does).

I doubt anyone is using the `providerInjector`, but still it might be a good idea to add `modules` on the `providerInjector` as well (for symmetry).

petebacondarwin replied:
I don't follow. You can't use `angular.module('myModule').info()` since you don't know if that module has been loaded or included in the current app.
In fact, you don't even know what the current app is since there might be more than one bootstrapped on a page.

Also, `$injector` is not ambiguous. It is the only injector that you can inject into components and services, etc.

petebacondarwin replied:
But maybe I'm missing the point...

dcherman replied:
I think he's referring to the `$injector` that you can inject into `config` function which is not the same injector as is injected into a service, component, etc.

gkalpak replied:
Yes, I am referring to the `providerInjector` that can be injected into providers/config blocks. E.g.:

```js
.provider('foo', ['$injector', function fooProvider(providerInjector) {
  this.$get = ['$injector', function foo(instanceInjector) {
    console.log(providerInjector === instanceInjector);   // false
  }
})
```

(Again, I don't expect many people to be actually using it, but you never know :smiley:)

--
Regarding `angular.module('someModule').info()`:
Assume you are the author of a library or addon that relies on another library (e.g. `myAwesomeMaterialToolbar`). You can request that your addon is loaded after `ngMaterial`, so you can detect its version and do different things. E.g.:

```js
// my-awesome-material-toolbar.js
var mdVersion = angular.module('ngMaterial').info().version;
var deps = (mdVersion >= '1.1.1') ? [] : ['myShimForAFeatureAddedInNgMaterial1.1.1'];

angular.module('myAwesomeMaterialToolbar', deps);
```

petebacondarwin replied:
I will fix the providerInjector/instanceInjector thing.

Regarding the load time dependencies use case. For sure you can do what you suggest with the functionality provided in this PR, although you are having to ensure that the developer loads the script files in the correct order and is not using some async loading helper that might mess things up.

It would be safer (IMO) to do configuration of your awesome library in providers, etc... Although I accept it is difficult to optionally add in a directive, you could make the directive a noop if not wanted.

gkalpak replied:
Works for me 👍

gkalpak commented on src/loader.js:
```diff
@@ -115,6 +118,39 @@ function setupModuleLoader(window) {
+           *                          or `this` if called as a setter.
+           *
+           * @description
+           * Additional info about this module
```
This makes it sound as if passing an object would extend the existing info (not overwrite it).
I think we should be pretty explicit about that.

petebacondarwin replied:
OK

gkalpak commented on test/loaderSpec.js:
```diff
@@ -156,4 +156,29 @@ describe('module loader', function() {
+
+      it('should store the object passed as a param', function() {
+        theModule.info({ version: '1.2' });
+        expect(theModule.info()).toEqual({ version: '1.2' });
```
I would add a test to verify it overwrites the previous info object.

petebacondarwin replied:
OK

petebacondarwin committed:
feat(Module): add `info()` method

The new `info()` method lets developers store arbitrary information about
their module for consumption later.

petebacondarwin committed:
feat($injector): add new `modules` property

The `modules` property is a hash of the modules loaded into the injector
at bootstrap time. This can be used to access the module's info.

petebacondarwin commented:
PTAL @gkalpak

gkalpak commented on src/loader.js:
```diff
@@ -115,6 +118,45 @@ function setupModuleLoader(window) {
+           */
+          info: function(value) {
+            if (isDefined(value)) {
+              if (!isObject(value)) throw ngMinErr('aobj', 'Argument \'{0}\' must be an object', 'value');
```
This should probably be `value` instead of `'value'` (i.e. without the quotes), right?

petebacondarwin replied:
No. I don't think so. The idea is that the error can be general and be used with other parameters names

gkalpak replied:
Right. Then why do you need the `{0}`? Why don't you just write `'Argument \'value\' must be an object'` directly? I guess in order to make it easier to recognize the part that is interchangable.
👍

gkalpak commented on test/auto/injectorSpec.js:
```diff
@@ -2,6 +2,37 @@
+      var test1 = angular.module('test1', ['test2']).info({ version: '1.1' });
+      var test2 = angular.module('test2', []).info({ version: '1.2' });
+      module('test1');
+      inject(['$injector', function($injector) {
```
OOC, is the array necessary?

petebacondarwin replied:
Didn't we decide to make all our tests minification proof?

petebacondarwin replied:
I'm sure there was an issue about it

gkalpak replied:
I don't remember anything like that (why would we)?
Maybe we have decided to make `ngMocks` minification-safe (because it makes it easier for some people to incorporate it in their workflow), but not our tests.

petebacondarwin replied:
Ah, yes you are right @gkalpak (as always) - I was thinking of this: https://github.com/angular/angular.js/issues/13542

mgol replied:
They don't have to be minification-proof but shouldn't we aim at making them pass in strict DI mode?

gkalpak replied:
There is not such thing as "strict DI" in tests :)
(Also none of the other tests are, so it is obviously not a problem.)

gkalpak commented:
LGTM <sub>(with or without the array notation :smiley:)</sub>

petebacondarwin commented:
Thanks @gkalpak - I just thought we should check that there is something similar in A2 before landing this.

petebacondarwin commented:
So in Angular there is the `NgModule` concept, which provides a similar place to store useful metadata about your modules/app. Let's merge this.

petebacondarwin committed:
feat(info): add `angularVersion` info to each module

You can now check what version of AngularJS a core module is designed for:

```
var angularVersion = $injector.modules['myModule'].info().angularVersion;
```

gkalpak commented on src/AngularPublic.js:
```diff
@@ -265,5 +265,6 @@ function publishExternalAPI(angular) {
     }
-  ]);
+  ])
+  .info({ angularVersion: '"NG_VERSION_FULL"' });
```
Why not call it just `version` (since it's already namespaced by the module name).
Especially in other modules (`ngAnimate`, `ngAria`, etc).

petebacondarwin replied:
I guess I was leaning towards the idea that you would stamp your module with the version of Angular that it was designed for, rather than the version of the module itself...

mgol replied:
To me `angularVersion` seems more intuitive; otherwise I'd feel we're talking about the module version as Pete said.

gkalpak replied:
I tend to think it as the module version indeed. But it doesn't make any real difference. `angularVersion` works for me.

gkalpak reviewed (commented):
Have you considered "version-stamping" `ngLocale`?

petebacondarwin commented:
I will add version stamping to the locale modules too.

petebacondarwin commented:
It turns out that adding the version stamp to the locale files is a little more involved with changes to grunt required :-(

gkalpak commented:
Sounds like it's not worth it. It still LGTM without vervion-stamping locale files :smiley:
